dora/core/common/src/test/java/alluxio/UfsUrlTest.java

```diff
@@ -118,13 +135,342 @@ public void hashCodeTest() {
+    UfsUrl ufsUrl1 = UfsUrl.createInstance("abc://1.2.3.4/testFolder1/testFolder2");
+    UfsUrl ufsUrl2 = UfsUrl.createInstance("abc://1.2.3.5/testFolder1");
+    UfsUrl ufsUrl3 = UfsUrl.createInstance("xyz://1.2.3.5/testFolder1/");
+    UfsUrl ufsUrl4 = UfsUrl.createInstance("abc://1.2.3.5/test");
+    UfsUrl ufsUrl5 = UfsUrl.createInstance("abc://1.2.3.4/testFolder1/");
+    UfsUrl ufsUrl6 = UfsUrl.createInstance("abc://1.2.3.5/testFolder1/testFolder2");
+
+    assertFalse(ufsUrl3.isAncestorOf(ufsUrl1));
+    assertFalse(ufsUrl2.isAncestorOf(ufsUrl1));
+    assertFalse(ufsUrl4.isAncestorOf(ufsUrl2));
+    assertFalse(ufsUrl3.isAncestorOf(ufsUrl6));
+    assertTrue(ufsUrl5.isAncestorOf(ufsUrl1));
+  }
+
+  /**
+   * Tests the {@link AlluxioURI#join(String)} and {@link AlluxioURI#join(AlluxioURI)} methods.
+   */
+  @Test
+  public void join() {
+    assertEquals(UfsUrl.createInstance("abc:/a"), UfsUrl.createInstance("abc:/").join("a"));
+    assertEquals(UfsUrl.createInstance("abc:/a/b.txt"),
+        UfsUrl.createInstance("abc:/a").join("/b.txt"));
+
+    final String pathWithSpecialChar = "根目录";
+    assertEquals(UfsUrl.createInstance("abc:/" + pathWithSpecialChar),
+        UfsUrl.createInstance("abc:/").join(pathWithSpecialChar));
+
+    final String pathWithSpecialCharAndColon = "根目录";
+    assertEquals(UfsUrl.createInstance("abc:/" + pathWithSpecialCharAndColon),
+        UfsUrl.createInstance("abc:/").join(pathWithSpecialCharAndColon));
+
+    // join empty string
+    assertEquals(UfsUrl.createInstance("abc:/a"), UfsUrl.createInstance("abc:/a").join(""));
+  }
+
+  /**
+   * Tests the {@link UfsUrl#createInstance(String)} constructor to work with file URIs
+   * appropriately.
+   */
+  @Test
+  public void fileUrl() {
+    UfsUrl url = UfsUrl.createInstance("file:///foo/bar");
+    assertTrue(url.getAuthority().toString().isEmpty());
+    assertEquals("/foo/bar", url.getFullPath());
+    assertEquals("file:///foo/bar", url.toString());
+  }
+
+  /**
+   * Tests the {@link UfsUrl#toString()} method to normalize paths.
+   */
+  @Test
+  public void normalize() {
+    assertEquals("xyz:///", UfsUrl.createInstance("xyz://").toString());
+    assertEquals("xyz:///foo", UfsUrl.createInstance("xyz:/foo/").toString());
+    assertEquals("xyz:///foo", UfsUrl.createInstance("xyz:/foo/").toString());
+    assertEquals("xyz:///foo", UfsUrl.createInstance("xyz:/foo/").toString());
+    assertEquals("xyz:///foo", UfsUrl.createInstance("xyz:/foo//").toString());
+    assertEquals("xyz:///foo/bar", UfsUrl.createInstance("xyz:/foo//bar").toString());
+
+    assertEquals("xyz:///foo/bar/boo", UfsUrl.createInstance("xyz:/foo/bar/boo").toString());
+    assertEquals("xyz:///foo/boo/baz",
+        UfsUrl.createInstance("xyz:/foo/////boo/baz").toString());
+    // TODO(Yichuan Sun): Is this case needed?
+//    assertEquals("xyz:///foo/boo", UfsUrl.createInstance("xyz:./foo/boo").toString());
+
+    assertEquals("foo://bar boo:8080/abc/c",
```

Comment:
let's mention in the comment in `UfsUrl` that for now we support whitespace chars in url